### PR TITLE
Laboratory work #2, Aleksandr Babii - 19FPL1

### PR DESCRIPTION
**Pull request by asbabiy**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
Push changes, please

File: lab_2/main.py [hunk 1]
```diff
@@ -34,7 +55,30 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
...
+    for check in checks:
+        if not check():
+            return []
+
+    lcs = create_zero_matrix(len(x), len(y))
+
+    for i, w1 in enumerate(x):
+        for j, w2 in enumerate(y):
+            if w1 == w2:
+                lcs[i][j] = lcs[i - 1][j - 1] + 1
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:80]: what happens if i == 0 or j == 0?

File: lab_2/main.py [hunk 2]
```diff
@@ -81,7 +198,32 @@ def calculate_text_plagiarism_score(original_text_tokens: tuple, suspicious_text
...
+    length_diff = len(origin) - len(suspicious)
+    plagiarism_sum = []
+
+    if length_diff < 0:
+        origin += tuple('' for _ in range(abs(length_diff)))
+        origin, suspicious = suspicious, origin
+
+    for sent_1, sent_2 in zip(origin, suspicious):
+        lcs_length = find_lcs_length(sent_1, sent_2, plagiarism_threshold)
+        plagiarism_sum.append(abs(calculate_plagiarism_score(lcs_length, sent_2)))
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: why do you need abs()?

File: lab_2/main.py [hunk 3]
```diff
@@ -108,7 +269,41 @@ def accumulate_diff_stats(original_text_tokens: tuple, suspicious_text_tokens: t
...
+        lcs_matrix = fill_lcs_matrix(sent_1, sent_2)
+        lcs = find_lcs(sent_1, sent_2, lcs_matrix)
+        plagiarism_score = calculate_plagiarism_score(lcs_length, sent_2)
+        diffs = find_diff_in_sentence(origin, suspicious, lcs)
+
+        sentence_plagiarism.append(plagiarism_score)
+        sentence_lcs_length.append(lcs_length)
+        difference_indexes.append(diffs)
+
+    diff_stats = {
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:283]: can you work directly with this dictionary?

File: lab_2/main.py [hunk 4]
```diff
@@ -119,7 +314,38 @@ def create_diff_report(original_text_tokens: tuple, suspicious_text_tokens: tupl
...
+
+        report_sentence = f'''
+            - {sent_1}
+            + {sent_2}
+
+            lcs = {lcs_length}, plagiarism = {plagiarism_score * 100}%
+
+            '''
+
+        print(report_sentence)
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: do not forget to add returns here

File: lab_2/main.py (hunk 2 above)
Comment (asbabiy) [reply lab_2/main.py]: Removed.

File: lab_2/main.py (hunk 3 above)
Comment (asbabiy) [reply lab_2/main.py:283]: Yep, fixed it.

File: lab_2/main.py (hunk 4 above)
Comment (asbabiy) [reply lab_2/main.py]: Ok

File: lab_2/main.py (hunk 1 above)
Comment (asbabiy) [reply lab_2/main.py:80]: Well, I got you. 

If i == 0 or j == 0, we start looking at the right end of the matrix, so indices in `lcs[i - 1][j -1] + 1` become negative (both or only of them, it doesn't really matter). This behavior can be handled using one more if condition but...

... 1) We start iterating over the first row and don't mutate other rows so we don't really bother about them even if we get lcs[-1][...], those values are all null at this point of time and become 1 if w1 == w2
2) Similar things happen with j == 0. Iterating over j and looking at it when it equals zero, we can notice that we mutate only the first column of the matrix, i.e. the first elements of each list so why should we care about the last ones? They are still unreachable for us at this moment.

Comment (ValeriyaKuznetsova) [reply lab_2/main.py:80]: Agree

File: lab_2/start.py [hunk 5]
```diff
@@ -1,4 +1,18 @@
...
+    origin = main.tokenize_big_file('lab_2/data.txt')
+    suspicious = main.tokenize_big_file('lab_2/data_2.txt')
+
+    lcs_length = main.find_lcs_length_optimized(origin, suspicious, plagiarism_threshold=0.3)
+    print(f"Longest common subsequence consists of {lcs_length} words.")
+    actual = main.calculate_plagiarism_score(lcs_length, suspicious)
+    RESULT = actual
+
+    # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
+    assert RESULT, 'Concordance not working'
```
Comment (ValeriyaKuznetsova) [inline lab_2/start.py]: RESULT == expected result

File: lab_2/main.py [hunk 6]
```diff
@@ -374,4 +388,17 @@ def tokenize_big_file(path_to_file: str) -> tuple:
     :param path_to_file: a path
     :return: a tuple with ids
     """
-    pass
+    tokens = []
+    idx = {}
+
+    file = (tokenize(row) for row in open(path_to_file, encoding='utf-8'))
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py]: the point is that you read a part of a file, tokenize it, transform into a numeric form and only then read another part
there should be only one large tuple stored (with ids), not two (with tokens and ids)

**Pull request closed without merging**